README.md
<p align="center"><a href="https://nodei.co/npm/discord-hybrid-sharding/"><img src="https://nodei.co/npm/discord-hybrid-sharding.png"></a></p>
<p align="center"><img src="https://img.shields.io/npm/v/discord-hybrid-sharding"> <img src="https://img.shields.io/npm/dm/discord-hybrid-sharding?label=downloads"> <img src="https://img.shields.io/npm/l/discord-hybrid-sharding"> <img src="https://img.shields.io/github/repo-size/meister03/discord-hybrid-sharding">  <a href="https://discord.gg/YTdNBHh"><img src="https://discordapp.com/api/guilds/697129454761410600/widget.png" alt="Discord server"/></a></p>

# Discord-Hybrid-Sharding

One first package which combines sharding manager & internal sharding to save a lot of resources, which allows clustering!

In other words: "Mixing both: if you need `x` shards for `n` process!"

**[NEW: TypeScript Rewrite | Upgrade Guide](https://gist.github.com/meister03/882ba6f6d805384f27336dd5ba389a54)**
**[NEW: Clustering Support for all JS Libraries](#Use-with-other-libraries)**

If you are interested in auto-scaling & cross-hosting on other machines, check out this package `npmjs.com/discord-cross-hosting`

### Featured by Discord Creators

[Private Community for Verified Bot Developers. ](https://discord.gg/R3hPevRtUV)
[Meet big bot and small bot developers and have a nice exchange...](https://discord.gg/R3hPevRtUV)

<p>
<a href="https://discord.gg/R3hPevRtUV">
<img src="https://media.discordapp.net/attachments/980770619161448489/982938274677018624/banner.png?width=320&height=80">
</a>
</p>

## Why?

The sharding manager is very heavy and uses more than 300MB per shard during light usage, while internal sharding uses just 20% of it. Internal sharding reaches its' limits at 14000 guilds and becomes slow when your bot gets bigger.

Your only solution becomes converting to the sharding manager. That's why this new package will solve all your problems (tested by many bots with 20-170k guilds), because it spawns shards, which has internal shards. **You can save up to 60% on resources!**

-   **Zero Downtime ReClustering/ReSharding/Restarts**
-   **Decentralized BroadCastEval function -> Listenerless, less memory leaks & cluster/client doesn't have to be ready**
-   **Heartbeat System -> Respawn unresponsive or dead `ClusterClient`s**
-   **IPC System -> Client <-> ClusterManager -> `.request()`, `.reply()`, `.send()`**
-   **Fine-grained control over the cluster queue -> `manager.queue.next(), .stop(), .resume()`**
-   Memory efficient -> 60% less memory when clustering
-   Debug event -> A good overview of cluster information
-   EvalOnManager function & other cool functions you need...
-   `Strings` & `Functions with Context` support on `.broadcastEval()`
-   Optional timeout feature on `.broadcastEval()` to prevent memory leaks
-   **[Supports cross-hosting: `Shard/Cluster` managing and cross-host communication (`.broadcastEval()`, `IPC`)](https://npmjs.com/discord-cross-hosting)**
-   **[Supports syncing Discord rate limits globally](https://npmjs.com/discord-cross-ratelimit)**
    **Scroll down to check our new functions!**

## How does it work?

There are clusters/master shards, which behave like regular shards in the sharding manager and spawns clusters in addition to internal shards. You do not have to spawn as many regular Shards (master shards), which can be replaced with internal shards.
"for process `n` , `n` internal shards"

Example: `A Discord bot with 4000 guilds`
Normally we would spawn 4 shards with the Sharding Manager (`~4 x 300MB memory`), but in this case we start with 2 clusters/master shards, which spawns 2 internal shards ==> We just saved 2 shards in comparison to the regular Sharding Manager (`~2 x 300MB memory`).

### See below for the Guide

**If you need help, feel free to join our <a href="https://discord.gg/YTdNBHh">Discord server</a>. ☺**

# Download

```cli
npm i discord-hybrid-sharding
------ or ---------------------
yarn add discord-hybrid-sharding
```

# Supports all Discord.js Versions & all other Libraries (Eris, Discordeno)

-   **Full Discord.js v12, v13, v14 support**
-   `Strings` and `Functions` with `context` are supported in `.broadcastEval()`
-   Most public methods accept sole objects, such as `.spawn({ amount: 20, timeout: -1 })`
-   Very similar functions to the Discord.js ShardingManager and more for the advanced usage

# Setting up

First, add the module into your project (into your shard/cluster file).
Filename: `Cluster.js`

```js
// Typescript: import { ClusterManager } from 'discord-hybrid-sharding'
const { ClusterManager } = require('discord-hybrid-sharding');

const manager = new ClusterManager(`${__dirname}/bot.js`, {
    totalShards: 7, // or 'auto'
    /// Check below for more options
    shardsPerClusters: 2,
    // totalClusters: 7,
    mode: 'process', // you can also choose "worker"
    token: 'YOUR_TOKEN',
});

manager.on('clusterCreate', cluster => console.log(`Launched Cluster ${cluster.id}`));
manager.spawn({ timeout: -1 });
```

After that, insert the code below into your `bot.js` file

```js
// Typescript: import { ClusterClient, getInfo } from 'discord-hybrid-sharding'
const { ClusterClient, getInfo } = require('discord-hybrid-sharding');
const Discord = require('discord.js');

const client = new Discord.Client({
    shards: getInfo().SHARD_LIST, // An array of shards that will get spawned
    shardCount: getInfo().TOTAL_SHARDS, // Total number of shards
});

client.cluster = new ClusterClient(client); // initialize the Client, so we access the .broadcastEval()
client.login('YOUR_TOKEN');
```

# Eval-ing over clusters

_Following examples assume that your `Discord.Client` is called `client`._

```js
client.cluster
    .broadcastEval(`this.guilds.cache.size`)
    .then(results => console.log(`${results.reduce((prev, val) => prev + val, 0)} total guilds`));

// or with a callback function
client.cluster
    .broadcastEval(c => c.guilds.cache.size)
    .then(results => console.log(`${results.reduce((prev, val) => prev + val, 0)} total guilds`));
```

# ClusterManager

| Option            | Type                  | Default  | Description                                                                                                                       |
| ----------------- | --------------------- | -------- | --------------------------------------------------------------------------------------------------------------------------------- |
| totalShards       | number or string      | "auto"   | Amount of internal shards which will be spawned                                                                                   |
| totalClusters     | number or string      | "auto"   | Amount of processes/clusters which will be spawned                                                                                |
| shardsPerClusters | number or string      | -        | Amount of shards which will be in one process/cluster                                                                             |
| shardList         | Array[number]         | -        | OPTIONAL - On cross-hosting or spawning specific shards you can provide a shardList of internal Shard IDs, which will get spawned |
| mode              | "worker" or "process" | "worker" | ClusterManager mode for the processes                                                                                             |
| token             | string                | -        | OPTIONAL -Bot token is only required totalShards are set to "auto"                                                                |

The Manager.spawn options are the same as for Sharding Manager

# Cluster Events

| Event         | Description                           |
| ------------- | ------------------------------------- |
| clusterCreate | Triggered when a Cluster gets spawned |

# Cluster Client Properties

All properties like for `.broadcastEval()` are available, just replace the `client.shard` with `client.cluster`
Other properties:
| Property | Description |
| ------------- | -------------- |
| client.cluster.count | Returns the amount of all clusters |
| client.cluster.id | Returns the current cluster ID |
| client.cluster.ids | Returns all internal shards of the cluster |

**Feel free to contribute/suggest or contact me on my Discord server or in DM: Meister#9667**

# Changes | Migrating to Discord-Hybrid-Sharding

Options are now labelled as `cluster` instead of `shard`:

```diff
- client.shard...
+ client.cluster...

- .broadcastEval((c, context) => c.guilds.cache.get(context.guildId), { context: { guildId: '1234' }, shard: 0 })
+ .broadcastEval((c, context) => c.guilds.cache.get(context.guildId), { context: { guildId: '1234' }, cluster: 0 })
```

Small changes in naming conventions:

```diff
- client.shard.respawnAll({ shardDelay = 5000, respawnDelay = 500, timeout = 30000 })
+ client.cluster.respawnAll({ clusterDelay: 5000, respawnDelay: 5500, timeout: 30000 })

- manager.shard.respawnAll({ shardDelay = 5000, respawnDelay = 500, timeout = 30000 })
+ manager.respawnAll({ clusterDelay: 5000, respawnDelay: 5500, timeout: 30000 })

```

Get current cluster ID:

```diff
- client.shard.id
+ client.cluster.id
```

Get current shard ID:

```diff
- client.shard.id
+ message.guild.shardId
```

Get total shards count:

```diff
- client.shard.count
+ client.cluster.info.TOTAL_SHARDS
```

Get all ShardID's in the current cluster:

```diff
- client.shard.id
+ [...client.cluster.ids.keys()]
```

# New functions & events:

## `Zero Downtime Reclustering`:

Zero Downtime Reclustering is a Plugin, which is used to reshard/recluster or even restart your bot with having a theoretical outage of some seconds.
There are two options for the `restartMode`:

-   `gracefulSwitch`: Spawns all new Clusters with the provided Info in maintenance mode, once all clusters have been spawned and the DiscordClient is ready, the clusters will exit maintenance mode, where as it will fire the `client.cluster.on('ready')` event. In order to load the Database and listen to events. Moreover all Clusters will be gracefully killed, once all clusters exited maintenance mode.
-   `rolling`: Spawns the Clusters with the provided Info in maintenance mode, once the DiscordClient is ready of the Cluster, the Cluster will exit maintenance mode, where as it will fire the `client.cluster.on('ready')` event. In order to load the Database and listen to events. Moreover the OldCluster will be killed, since the Cluster has exited maintenance mode. Not recommended, when shardData has not been updated.

Cluster.js

```js
// Typescript: import { ClusterManager, ReClusterManager  } from 'discord-hybrid-sharding'
const { ClusterManager, ReClusterManager } = require('discord-hybrid-sharding');
const manager = new ClusterManager(`${__dirname}/bot.js`, {...});

manager.extend(
    new ReClusterManager()
)
... ///SOME CODE
// Start reclustering
const optional = {totalShards, totalClusters....}
manager.recluster?.start({restartMode: 'gracefulSwitch', ...optional})
```

Bot.js

```js
// Typescript: import { ClusterClient, getInfo } from 'discord-hybrid-sharding'
const { ClusterClient, getInfo } = require('discord-hybrid-sharding');
const client = new Discord.Client(...)
client.cluster = new Cluster.Client(client);

if (client.cluster.maintenance) console.log(`Bot on maintenance mode with ${client.cluster.maintenance}`);

client.cluster.on('ready', () => {
    // Load Events
    // Handle Database stuff, to not process outdated data
});

client.login(token);
```

## `HeartbeatSystem`

-   Checks if Cluster/Client sends a heartbeat on a given interval
-   When the Client doesn't send a heartbeat, it will be marked as dead/unresponsive
-   Cluster will get respawned after the given amount of missed heartbeats has been reached

```js
// Typescript: import { ClusterManager, HeartbeatManager  } from 'discord-hybrid-sharding'
const { ClusterManager, HeartbeatManager } = require('discord-hybrid-sharding');
const manager = new ClusterManager(`${__dirname}/bot.js`, {...});

manager.extend(
    new HeartbeatManager({
        interval: 2000, // Interval to send a heartbeat
        maxMissedHeartbeats: 5, // Maximum amount of missed Heartbeats until Cluster will get respawned
    })
)
```

## `Control Restarts`

-   Cap the amount of restarts per cluster to a given amount on a given interval

```js
const manager = new ClusterManager(`${__dirname}/bot.js`, {
    ...YourOptions,
    restarts: {
        max: 5, // Maximum amount of restarts per cluster
        interval: 60000 * 60, // Interval to reset restarts
    },
});
```

## `IPC System`

-   The IPC System allows you to listen to your messages
-   You can communicate between the cluster and the client
-   This allows you to send requests from the client to the cluster and reply to them and vice versa
-   You can also send normal messages which do not need to be replied

ClusterManager | `cluster.js`

```js
// Typescript: import { ClusterManager, messageType } from 'discord-hybrid-sharding'
const { ClusterManager, messageType } = require('discord-hybrid-sharding');
const manager = new ClusterManager(`${__dirname}/testbot.js`, {
    totalShards: 1,
    totalClusters: 1,
});

manager.on('clusterCreate', cluster => {
    cluster.on('message', message => {
        console.log(message);
        if (message._type !== messageType.CUSTOM_REQUEST) return; // Check if the message needs a reply
        message.reply({ content: 'hello world' });
    });
    setInterval(() => {
        cluster.send({ content: 'I am alive' }); // Send a message to the client
        cluster.request({ content: 'Are you alive?', alive: true }).then(e => console.log(e)); // Send a message to the client
    }, 5000);
});
manager.spawn({ timeout: -1 });
```

ClusterClient | `client.js`

```js
// Typescript: import { ClusterClient, getInfo, messageType } from 'discord-hybrid-sharding'

changes.md
# Changes

```diff
- ShardClientUtil.js
+ ClusterClient.js:
   -> Change Shard to Cluster
   -> Add Info Function, which parses out Process.Env or Worker Data
   -> Remove Broadcast Logic of Discord.js
   -> Convert to Typescript
- Shard.js
+ Cluster.js
   -> Change Shard to Cluster
   -> Add new Error Event
   -> Change Spawn Options
   -> Convert to Typescript
- ShardManager.js
+ ClusterManager.js
    --> Change Shard to Cluster
    --> Add Chunk Function
    --> Add Option to Accept Internal Shard List
    --> Add private Properties
    -> Convert to Typescript

+ Util.js
   --> Copy Discord.js Util Functions for making Library zero Dep
   -> Convert to Typescript

+ Util.js
   --> Copy Discord.js Constants for making Library zero Dep
   -> Convert to Typescript

+ interface ClusterManager, Cluster, ClusterClient
  --> Handy Event Emitter Typings from @discordjs/rest
```

**Every file, which is not listed above follows the License of the Repo**


src/Cluster.js
const EventEmitter = require('events');
const path = require('path');
const Util = require('./Util.js');
const { IPCMessage, BaseMessage } = require('./IPCMessage.js')
let childProcess = null;
let Worker = null;

/**
 * A self-contained cluster created by the {@link ClusterManager}. Each one has a {@link ChildProcess} that contains
 * an instance of the bot and its {@link Client}. When its child process/worker exits for any reason, the cluster will
 * spawn a new one to replace it as necessary.
 * @extends EventEmitter
 */
class Cluster extends EventEmitter {
  /**
   * @param {ClusterManager} manager Manager that is creating this cluster
   * @param {number} id ID of this cluster
   */
  constructor(manager, id, shardlist, totalshards) {
    super();
    if (manager.mode === 'process') childProcess = require('child_process');
    else if (manager.mode === 'worker') Worker = require('worker_threads').Worker;
    /**
     * Manager that created the cluster
     * @type {ClusterManager}
     */
    this.manager = manager;
    /**
     * ID of the cluster in the manager
     * @type {number}
     */
    this.id = id;

    /**
    * Arguments for the shard's process (only when {@link ShardingManager#mode} is `process`)
    * @type {string[]}
    */
    this.args = manager.shardArgs || [];

    /**
     * Arguments for the shard's process executable (only when {@link ShardingManager#mode} is `process`)
     * @type {string[]}
     */
    this.execArgv = manager.execArgv;
    /**
    * Internal Shards which will get spawned in the cluster
    * @type {number}
    */
    this.shardlist = shardlist;
    /**
    * the amount of real shards
    * @type {number}
    */
    this.totalshards = totalshards;
    /**
    * Environment variables for the cluster's process, or workerData for the cluster's worker
    * @type {Object}
    */
    this.env = Object.assign({}, process.env, {
      SHARD_LIST: this.shardlist,
      TOTAL_SHARDS: this.totalshards,
      CLUSTER_MANAGER: true,
      CLUSTER: this.id,
      CLUSTER_COUNT: this.manager.totalClusters,
      KEEP_ALIVE_INTERVAL: this.manager.keepAlive.interval,
      DISCORD_TOKEN: this.manager.token,
    });
    /**
    * Whether the cluster's {@link Client} is ready
    * @type {boolean}
    */

    /**
     * Process of the cluster (if {@link ClusterManager#mode} is `process`)
     * @type {?ChildProcess}
     */
    this.process = null;

    /**
     * Worker of the cluster (if {@link ClusterManager#mode} is `worker`)
     * @type {?Worker}
     */
    this.worker = null;

    /**
    * The Hearbeat Object, which contains the missed Hearbeats, the last Hearbeat and the Hearbeat Interval
    * @type {Object}
    */
    this.heartbeat = {};


    /**
    * The Amount of maximal Restarts, which can be executed by the HeartBeat Checkup
    * @type {Object}
    */
    this._restarts = {};
    if (this.manager.keepAlive) {
      if (Object.keys(this.manager.keepAlive)) {
        this._restarts.current = 0;
        this._restarts.interval = setInterval(() => {
          this._restarts.current = 0;
        }, 60000 * 60);
      }
    }


    /**
     * Ongoing promises for calls to {@link Cluster#eval}, mapped by the `script` they were called with
     * @type {Map<string, Promise>}
     * @private
     */
    this._evals = new Map();

    /**
     * Ongoing promises for calls to {@link Cluster#fetchClientValue}, mapped by the `prop` they were called with
     * @type {Map<string, Promise>}
     * @private
     */
    this._fetches = new Map();



    /**
     * Listener function for the {@link ChildProcess}' `exit` event
     * @type {Function}
     * @private
     */
    this._exitListener = this._handleExit.bind(this, undefined);



  }
  /**
 * Forks a child process or creates a worker thread for the cluster.
 * <warn>You should not need to call this manually.</warn>
 * @param {number} [spawnTimeout=30000] The amount in milliseconds to wait until the {@link Client} has become ready
 * before resolving. (-1 or Infinity for no wait)
 * @returns {Promise<ChildProcess>}
 */
  async spawn(spawnTimeout = 30000) {
    if (this.process) throw new Error('CLUSTERING_PROCESS_EXISTS', this.id);
    if (this.worker) throw new Error('CLUSTERING_WORKER_EXISTS', this.id);
    if (this.manager.mode === 'process') {
      this.process = childProcess
        .fork(path.resolve(this.manager.file), this.args, {
          env: this.env,
          execArgv: this.execArgv,
        })
        .on('message', this._handleMessage.bind(this))
        .on('exit', this._exitListener)
        .on('error', this._handleError.bind(this));
    } else if (this.manager.mode === 'worker') {
      this.worker = new Worker(path.resolve(this.manager.file), { workerData: this.env })
        .on('message', this._handleMessage.bind(this))
        .on('exit', this._exitListener)
        .on('error', this._handleError.bind(this));

    }


    this._evals.clear();
    this._fetches.clear();

    /**
     * Emitted upon the creation of the cluster's child process/worker.
     * @event Cluster#spawn
     * @param {ChildProcess|Worker} process Child process/worker that was created
     */
    this.emit('spawn', this.process || this.worker);

    if (spawnTimeout === -1 || spawnTimeout === Infinity) return this.process || this.worker;
    spawnTimeout = spawnTimeout * this.shardlist.length;
    await new Promise((resolve, reject) => {
      const cleanup = () => {
        clearTimeout(spawnTimeoutTimer);
        this.off('ready', onReady);
        this.off('disconnect', onDisconnect);
        this.off('death', onDeath);
      };

      const onReady = () => {
        this._cleanupHearbeat()
        cleanup();
        resolve();
      };

      const onDisconnect = () => {
        cleanup();
        reject(new Error('CLUSTERING_READY_DISCONNECTED', this.id));
      };

      const onDeath = () => {
        cleanup();
        reject(new Error('CLUSTERING_READY_DIED', this.id));
      };

      const onTimeout = () => {
        cleanup();
        reject(new Error('CLUSTERING_READY_TIMEOUT', this.id));
      };

      const spawnTimeoutTimer = setTimeout(onTimeout, spawnTimeout);
      this.once('ready', onReady);
      this.once('disconnect', onDisconnect);
      this.once('death', onDeath);
    });
    return this.process || this.worker;
  }
  /**
  * Immediately kills the clusters's process/worker and does not restart it.
  * @param {Object} options Some Options for managing the Kill
  * @param {Object} options.force Whether the Cluster should be force kill and be ever respawned...
  */
  kill(options = {}) {
    if (this.process) {
      try{ 
        this.process.removeListener('exit', this._exitListener);
      }catch(error){}
      this.process.kill();
    } else {
      try{
        this.worker.removeListener('exit', this._exitListener);
      }catch(error){}
      this.worker.terminate();
    }
    if (options.force) this._cleanupHearbeat()
    this._handleExit(false);
  }
  /**
  * Kills and restarts the cluster's process/worker.
  * @param {ClusterRespawnOptions} [options] Options for respawning the cluster
  * @returns {Promise<ChildProcess>}
  */
  async respawn({ delay = 500, timeout = 30000 } = {}) {
    if (this.worker || this.process) this.kill({ force: true });
    if (delay > 0) await Util.delayFor(delay);
    return this.spawn(timeout);
  }
  /**
  * Sends a message to the cluster's process/worker.
  * @param {*|BaseMessage} message Message to send to the cluster
  * @returns {Promise<Shard>}
  */
  send(message) {
    if (typeof message === 'object') message = new BaseMessage(message).toJSON();
    return new Promise((resolve, reject) => {
      if (this.process) {
        this.process.send(message, err => {
          if (err) reject(err);
          else resolve(this);
        });
      } else {
        if (!this.worker) return resolve({ err: 'WORKER_DOES_NOT_EXIST' });
        this.worker.postMessage(message);
        resolve(this);
      }
    });
  }

  /**
  * Sends a Request to the ClusterClient and returns the reply
  * @param {BaseMessage} message Message, which should be sent as request
  * @returns {Promise<*>} Reply of the Message
  * @example
  * client.cluster.request({content: 'hello'})
  *   .then(result => console.log(result)) //hi
  *   .catch(console.error);
  * @see {@link IPCMessage#reply}
  */
  request(message = {}) {
    message._sRequest = true;
    message._sReply = false;
    message = new BaseMessage(message).toJSON()
    return new Promise((resolve, reject) => {
      const nonce = message.nonce;
      this.manager._nonce.set(nonce, { resolve, reject });
      const sent = this.send(message, void 0, void 0, e => {
        if (e) reject(new Error(`Failed to deliver Message to cluster: ${e}`));
        setTimeout(() => {
          if (this.manager._nonce.has(nonce)) {
            this.manager._nonce.get(nonce).reject(new Error("Eval Request Timedout"));
            this.manager._nonce.delete(nonce);
          }
        }, (message.timeout || 10000));
      });
      if (!sent) reject(new Error("CLUSTERING_IN_PROCESS or CLUSTERING_DIED"));
    }).catch((e) => ({ ...message, error: new Error(e.toString()) }))
  }

  /**
  * Fetches a client property value of the cluster.
  * @param {string} prop Name of the client property to get, using periods for nesting
  * @returns {Promise<*>}
  * @example
  * cluster.fetchClientValue('guilds.cache.size')
  *   .then(count => console.log(`${count} guilds in cluster ${cluster.id}`))
  *   .catch(console.error);
  */
  fetchClientValue(prop) {
    // Shard is dead (maybe respawning), don't cache anything and error immediately
    if (!this.process && !this.worker) return Promise.reject(new Error('CLUSTERING_NO_CHILD_EXISTS', this.id));

    // Cached promise from previous call
    if (this._fetches.has(prop)) return this._fetches.get(prop);

    const promise = new Promise((resolve, reject) => {
      const child = this.process || this.worker;

      const listener = message => {
        if (!message || message._fetchProp !== prop) return;
        child.removeListener('message', listener);
        this._fetches.delete(prop);
        resolve(message._result);
      };
      child.on('message', listener);

      this.send({ _fetchProp: prop }).catch(err => {
        child.removeListener('message', listener);
        this._fetches.delete(prop);
        reject(err);
      });
    });

    this._fetches.set(prop, promise);
    return promise;
  }


  /**
  * Evaluates a script or function on the cluster, in the context of the {@link Client}.
  * @param {string|Function} script JavaScript to run on the cluster
  * @returns {Promise<*>} Result of the script execution
  */
  eval(script, context, timeout) {
    // Stringify the script if it's a Function
    const _eval = typeof script === 'function' ? `(${script})(this, ${JSON.stringify(context)})` : script;

    // cluster is dead (maybe respawning), don't cache anything and error immediately
    if (!this.process && !this.worker) return Promise.reject(new Error('CLUSTERING_NO_CHILD_EXISTS', this.id));

    // Cached promise from previous call
    if (this._evals.has(_eval)) return this._evals.get(_eval);

    const promise = new Promise((resolve, reject) => {
      const child = this.process || this.worker;

      //Set A timeout for the eval
      let temptimeout;
      
      const listener = message => {
        if (!message || message._eval !== _eval) return;
        if(temptimeout) clearTimeout(temptimeout);
        child.removeListener('message', listener);
        this._evals.delete(_eval);
        if (!message._error) resolve(message._result);
        else reject(Util.makeError(message._error));
      };
      child.on('message', listener);

      this.send({ _eval }).then(m => {
        if(timeout){
          temptimeout = setTimeout(()=> {
            if(!this._evals.has(_eval)) return;
            child.removeListener('message', listener);
            this._evals.delete(_eval);
            reject(new Error(`BROADCAST_EVAL_REQUEST_TIMED_OUT`));
          }, timeout)
        }
      }).catch(err => {
        if(temptimeout) clearTimeout(temptimeout);
        child.removeListener('message', listener);
        this._evals.delete(_eval);
        reject(err);
      });
    });

    this._evals.set(_eval, promise);
    return promise;
  }

  /**
   * Handles a message received from the child process/worker.
   * @param {*} message Message received
   * @private
   */
  _handleMessage(message) {
    if (message) {
      // Cluster is ready
      if (message._ready) {
        this.ready = true;
        /**
         * Emitted upon the cluster's {@link Client#ready} event.
         * @event Cluster#ready
         */
        this.emit('ready');
        this.manager._debug('Ready', this.id);
        this._cleanupHearbeat()
        this._checkIfClusterAlive()
        return;
      }

      // Cluster has disconnected
      if (message._disconnect) {
        this.ready = false;
        /**
         * Emitted upon the cluster's {@link Client#disconnect} event.
         * @event Cluster#disconnect
         */
        this.emit('disconnect');
        this.manager._debug('[DISCONNECT] Some Shards disconnected', this.id);
        return;
      }

      // Cluster is attempting to reconnect
      if (message._reconnecting) {
        this.ready = false;
        /**
         * Emitted upon the cluster's {@link Client#reconnecting} event.
         * @event Cluster#reconnecting
         */
        this.emit('reconnecting');
        this.manager._debug('[RECONNECTING] Some Shards are attempting reconnect', this.id);
        return;
      }

      //The Hearbeat, which has been sent by Client
      if (message._keepAlive) {
        if (this.manager.keepAlive) this._heartbeatMessage(message);
        return;
      }

      // Cluster is requesting a property fetch
      if (message._sFetchProp) {
        const resp = { _sFetchProp: message._sFetchProp, _sFetchPropShard: message._sFetchPropShard };
        this.manager.fetchClientValues(message._sFetchProp, message._sFetchPropShard).then(
          results => this.send({ ...resp, _result: results }),
          err => this.send({ ...resp, _error: Util.makePlainError(err) }),
        );
        return;
      }

      // Cluster is requesting an eval broadcast
      if (message._sEval) {
        const resp = { _sEval: message._sEval, _sEvalShard: message._sEvalShard };
        this.manager._performOnShards('eval', [message._sEval], message._sEvalShard, message._sEvalTimeout).then(
          results => this.send({ ...resp, _result: results }),
          err => this.send({ ...resp, _error: Util.makePlainError(err) }),
        );
        return;
      }

      //Evals a Request on a Cluster
      if (message.hasOwnProperty('_sManagerEval')) {
        this.manager.evalOnManager(message._sManagerEval).then((result) => {
          if (result._results) return this.send({ _sManagerEvalResponse: result._results, ...message });
          if (result._error) return this.send({ _error: Util.makePlainError(result._error), ...message })
        });
        return;
      }

      //Evals a Request on a Cluster
      if (message.hasOwnProperty('_sClusterEval')) {
        this.manager.evalOnCluster(message._sClusterEval, { ...message, requestcluster: this.id }).catch((e) => new Error(e))
        return;
      }

      //If Message is a Eval Response
      if (message.hasOwnProperty('_sClusterEvalResponse')) {
        const promise = this.manager._nonce.get(message.nonce);
        if (!promise) return;
        if (promise) {
          if (message._error) {
            promise.reject(message._error)
            this.manager._nonce.delete(message.nonce);
            if (this.manager.clusters.has(promise.requestcluster)) {
              this.manager.clusters.get(promise.requestcluster).send(message)
            }
          } else {
            promise.resolve(message._sClusterEvalResponse)
            this.manager._nonce.delete(message.nonce);
            if (this.manager.clusters.has(promise.requestcluster)) {
              this.manager.clusters.get(promise.requestcluster).send(message)
            }
          }
        }
        return;
      }

      // Cluster is requesting a respawn of all shards
      if (message._sRespawnAll) {
        const { clusterDelay, respawnDelay, timeout } = message._sRespawnAll;
        this.manager._debug('Cluster requested respawn of all Clusters', this.id);
        this.manager.respawnAll({ clusterDelay, respawnDelay, timeout }).catch(() => {
          // Do nothing
        });
        return;
      }

      if (message._sCustom) {
        if (message._sReply) {
          const promise = this.manager._nonce.get(message.nonce);
          if (promise) {
            promise.resolve(message)
            this.manager._nonce.delete(message.nonce);
          }
        } else if (message._sRequest) {
          // this.manager.request(message).then(e => this.send(e)).catch(e => this.send({...message, error: e}))
        }
      }
    }

    let emitmessage;
    if (typeof message === 'object'){
       emitmessage = new IPCMessage(this, message);
       if(emitmessage._sRequest) this.manager.emit('clientRequest', emitmessage)
    }else emitmessage = message;
    /**
    * Emitted upon receiving a message from the child process/worker.
    * @event Shard#message
    * @param {*|IPCMessage} message Message that was received
    */
    this.emit('message', emitmessage);
  }

  /**
   * Handles the cluster's process/worker exiting.
   * @param {boolean} [respawn=this.manager.respawn] Whether to spawn the cluster again
   * @private
   */
  _handleExit(respawn = this.manager.respawn) {
    /**
     * Emitted upon the cluster's child process/worker exiting.
     * @event Cluster#death
     * @param {ChildProcess|Worker} process Child process/worker that exited
     */
    this.emit('death', this.process || this.worker);
    this.manager._debug('[DEATH] Cluster died, attempting respawn', this.id);

    this.ready = false;
    this.process = null;
    this.worker = null;
    this._evals.clear();
    this._fetches.clear();

    if (respawn) this.spawn().catch(err => this.emit('error', err));
  }

  ///Custom Functions

  /**
   * Handles the cluster's process/worker error.
   * @param {Object} [error] the error, which occured on the worker/child process
   * @private
   */
  _handleError(error) {
    /**
    * Emitted upon the cluster's child process/worker error.
    * @event Cluster#error
    * @param {ChildProcess|Worker} process Child process/worker, where error occured
    */
    this.manager.emit('error', error);
  }


  /**
  * Handles the keepAlive Heartbeat and validates it.
  * @param {Object} [message] the heartbeat message, which has been recieved
  * @private
  */
  _heartbeatMessage(message) {
    if (!this.manager.keepAlive) return;
    if (Object.keys(this.manager.keepAlive).length === 0) return;
    this.heartbeat.last = Date.now();
    this.heartbeat.missed = 0;
    this.send({ ack: true, last: Date.now() }).catch(e => this.manager._debug('[ACK_FAILED] ACK could not be sent. IPC Channel is dead or busy', this.id))
  }


  _checkIfClusterAlive() {
    if (!this.manager.keepAlive) return;
    if (Object.keys(this.manager.keepAlive).length === 0) return;
    this.manager._debug('Hearbeat Interval CheckUp has started', this.id);
    this.heartbeat.interval = setInterval(() => {
      if (!this.heartbeat) return;
      const diff = Date.now() - Number(this.heartbeat.last);
      if (isNaN(diff)) return;
      if (diff > (this.manager.keepAlive.interval + 2000)) {
        this.heartbeat.missed++
        if (this.heartbeat.missed < this.manager.keepAlive.maxMissedHeartbeats) this.manager._debug(`[Heartbeat_MISSING] ${this.heartbeat.missed} Heartbeat(s) have been missed.`, this.id);
        if (this.heartbeat.missed > this.manager.keepAlive.maxMissedHeartbeats) {
          if (this._restarts) {
            if (this._restarts.current !== undefined) {
              this._restarts.current++
              if (this._restarts.current === this.manager.keepAlive.maxClusterRestarts) {
                return this.manager._debug(`[Heartbeat_MISSING] Already attempted maximal Amount of Respawns in less then 1 hour | Cluster will be respawned, when the cooldown ends.`, this.id);
              }
              if (this._restarts.current > this.manager.keepAlive.maxClusterRestarts) return;
            }
          }
          this.manager._debug(`[Heartbeat_MISSING] Attempting respawn | To much hearbeats were missing.`, this.id);
          this._cleanupHearbeat()
          this.respawn()
        }
      } else return;
    }, this.manager.keepAlive.interval);
    return this.heartbeat;
  }

  _cleanupHearbeat() {
    clearInterval(this.heartbeat.interval);
    if (Object.keys(this.manager.keepAlive).length === 0) return;
    this.heartbeat = {};
    this.manager._debug('[Heartbeat] Heartbeat has been cleared', this.id);
    return this.heartbeat;
  }
}

module.exports = Cluster;
